src/jdk.security.auth/share/classes/com/sun/security/auth/module/GssLoginModule.java

```diff
@@ -0,0 +1,656 @@
+                    doNotPrompt ? "true" : "false"));
+        if (!initiate && !accept)
+            initiate = true;
+        if (debug) {
+            System.out.print("Debug is  " + debug
+                             + " doNotPrompt " + doNotPrompt
+                             + " defName is " + defName
+                             + " nametype is " + nametype
+                             + " tryFirstPass is " + tryFirstPass
+                             + " useFirstPass is " + useFirstPass
+                             + " storePass is " + storePass
+                             + " clearPass is " + clearPass
+                             + " initiate is " + initiate
+                             + " accept is " + accept
+                             + " tryDefaultCreds is " + tryDefaultCreds
+                             + " useDefaultCreds is " + useDefaultCreds + "\n");
+        }
+    }
+
+
+    /**
+     * Authenticate the user
+     *
+     * <p>
+     *
+     * @return true in all cases since this <code>LoginModule</code>
+     *          should not be ignored.</p>
+     *
+     * <p>
+     * @exception FailedLoginException if the authentication fails.</p>
+     *
+     * <p>
+     * @exception LoginException if this <code>LoginModule</code>
+     *          is unable to perform the authentication.</p>
+     */
+    public boolean login() throws LoginException {
+        succeeded = false;
+
+        /*
+         * See commentary in initialize().  By returning false we cause
+         * LoginContext to ignore this module.
+         */
+        if (!useNative)
+            return false;
+        try {
+            if (tryFirstPass || useFirstPass) {
+                attemptAuthentication(true);
+                if (debug)
+                    System.out.println("\t\t[GssLoginModule] " +
+                                       "authentication succeeded");
+                succeeded = true;
+                cleanState();
+                return true;
+            }
+        } catch (LoginException le) {
+            // authentication failed -- try again below by prompting
+            cleanState();
+            if (debug) {
+                System.out.println("\t\t[GssLoginModule] " +
+                                   (tryFirstPass ? "tryFirstPass " : "") +
+                                   "authentication failed with:" +
+                                   le.getMessage());
+            }
+            if (useFirstPass)
+                throw le;
+        }
+
+        // The first password didn't work or we didn't try it, try prompting
+        try {
+            attemptAuthentication(false);
+            if (debug)
+                System.out.println("\t\t[GssLoginModule] " +
+                                   "authentication succeeded");
+            succeeded = true;
+            cleanState();
+            return true;
+        } catch (LoginException le2) {
+            cleanState();
```

Comment:
This should be in a `finally` block, I think.

Reply:
Agreed!  Again, a symptom of `Krb5LoginModule` being ancient?  But we should modernize `GssLoginModule` for sure.